projects/plugin/data/example.html

```diff
@@ -1,47 +1,244 @@
+        144: 'num lock',
+        145: 'scroll lock',
+        160: 'l shift',
+        161: 'r shift',
+        162: 'l ctrl',
+        163: 'r ctrl',
+        164: 'l alt',
+        165: 'r alt',
+        186: ';',
+        187: '=',
+        188: ',',
+        189: '-',
+        190: '.',
+        191: '/',
+        192: '`',
+        219: '[',
+        220: '\\',
+        221: ']',
+        222: '\'',
+    }
+
+    // History
+    var _history = document.getElementById("history");
+    var _history_curr_pressed = new Set();
+    var _history_key_pressing = false
+    var _history_max = 50;
+
+    function onKeyEvent(data) {
+        if (data.event_type.startsWith('mouse')) return;
+
+        if (data.event_type === "key_pressed") {
+            _history_curr_pressed.add(data.rawcode)
+            _history_key_pressing = true
+        } else if (data.event_type === "key_released") {
+            _history_curr_pressed.delete(data.rawcode)
+            _history_key_pressing = false
+        }
+
+        updateUI()
+    }
+
+    function updateUI() {
+        var _has_keys = _history_curr_pressed.size > 0;
+        var _has_children = _history.children.length > 0;
+
+        if (!_has_children && _history_key_pressing) {
+            _history.insertAdjacentHTML("beforeend", `<li></li>`)
+        }
+
+        if (_history_key_pressing) {
+            _history.lastElementChild.innerHTML = concatKeysFromSet(_history_curr_pressed)
+        } else {
+            if (_history.lastElementChild.innerHTML !== "") {
+                _history.insertAdjacentHTML("beforeend", `<li></li>`)
             }
         }
 
-        function start_websocket() {
-            var ws = new WebSocket("ws://localhost:16899/");
+        // Remove first element
+        if (_history.children.length > _history_max) _history.firstChild.remove();
+
+        // Scroll down
+        scrollTo({
+            top: document.body.scrollHeight
+        })
+    }
+
+    function concatKeysFromSet(_set_keycodes) {
+        return Array.from(_set_keycodes).map(code => KEYCODES[code] ?? '').join(" + ");
+    }
+
+    // Logger
+    var _log = document.getElementById("log");
+    var _log_require_clear = false;
+    var _log_count = 0;
 
-            ws.onmessage = on_data;
-            ws.onmessage = on_data;
-            ws.onerror = (e) => console.log('WebSocket error: ' + e);
+    function logData(data) {
+        _log.insertAdjacentHTML("beforeend", `<li>${data}</li>`);
 
-            ws.onclose = () => {
-                // connection closed, discard old websocket and create a new one in 5s
-                ws = null;
-                setTimeout(startWebsocket, 2000);
-            };
+        // Only keep last 50 entries
+        if (!_log_require_clear) {
+            _log_count++;
+            if (_log_count > 50) _log_require_clear = true;
+        } else {
+            _log.children[0].remove();
         }
-        
-        start_websocket();
-    </script>
-</html>
+    }
```

Comment:
If we move input history into a separate html file you can get rid of the logging since that's just for the example

Reply:
Where should I put this separate HTML file?